OrbitVulkanLayer/SubmissionTracker.h

```diff
@@ -0,0 +1,816 @@
+  // markers into the `QueueSubmission`, to make it persistent across submissions and such that it
+  // can be picked up later (on a vkQueuePresentKHR) to retrieve the timer results and send the data
+  // to the client.
+  //
+  // We assume to be capturing if `queue_submission_optional` has content, i.e. we were capturing
+  // on this VkQueueSubmit before calling into the driver.
+  // The meta information allows allows us to map submissions from the Vulkan layer to the driver
+  // submissions.
+  void PersistDebugMarkersOnSubmit(VkQueue queue, uint32_t submit_count,
+                                   const VkSubmitInfo* submits,
+                                   std::optional<QueueSubmission> queue_submission_optional) {
+    absl::WriterMutexLock lock(&mutex_);
+    if (!queue_to_markers_.contains(queue)) {
+      queue_to_markers_[queue] = {};
+    }
+    CHECK(queue_to_markers_.contains(queue));
+    QueueMarkerState& markers = queue_to_markers_.at(queue);
+
+    // If we consider that we are still capturing, take a cpu timestamp as "post submission" such
+    // that the submission "meta information" is complete. We can then attach that also to each
+    // debug marker, as they may be spanned across different submissions.
+    if (queue_submission_optional.has_value()) {
+      queue_submission_optional->meta_information.post_submission_cpu_timestamp =
+          MonotonicTimestampNs();
+    }
+
+    std::vector<uint32_t> marker_slots_to_reset;
+    VkDevice device = VK_NULL_HANDLE;
+    for (uint32_t submit_index = 0; submit_index < submit_count; ++submit_index) {
+      VkSubmitInfo submit_info = submits[submit_index];
+      for (uint32_t command_buffer_index = 0; command_buffer_index < submit_info.commandBufferCount;
+           ++command_buffer_index) {
+        VkCommandBuffer command_buffer = submit_info.pCommandBuffers[command_buffer_index];
+        if (device == VK_NULL_HANDLE) {
+          CHECK(command_buffer_to_device_.contains(command_buffer));
+          device = command_buffer_to_device_.at(command_buffer);
+        }
+        CHECK(command_buffer_to_state_.contains(command_buffer));
+        CommandBufferState& state = command_buffer_to_state_.at(command_buffer);
+
+        for (const Marker& marker : state.markers) {
+          std::optional<SubmittedMarker> submitted_marker = std::nullopt;
+          if (marker.slot_index.has_value() && queue_submission_optional.has_value()) {
+            submitted_marker = {.meta_information = queue_submission_optional->meta_information,
+                                .slot_index = marker.slot_index.value()};
+          }
+
+          switch (marker.type) {
+            case MarkerType::kDebugMarkerBegin: {
+              if (queue_submission_optional.has_value() && marker.slot_index.has_value()) {
+                ++queue_submission_optional->num_begin_markers;
+              }
+              CHECK(marker.label_name.has_value());
```

Comment:
Random note to be discussed later: We should check performance overhead of all these CHECKs here. (I think we talked about this before, though I don't recall what the conclusion was.)

Reply:
Agree.